blockchain/validators/src/metadata_key.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2018-20&&22 The MobileCoin Foundation
+
+//! Block metadata key validator.
+
+use crate::{Error, NodeKeyValidity};
+use mc_blockchain_types::{BlockIndex, BlockMetadata};
+use mc_crypto_keys::Ed25519Public;
+use std::collections::{BTreeMap, BTreeSet};
+
+/// A validator for block metadata signing keys and signatures.
+#[derive(Clone, Debug)]
+pub struct MetadataKeyValidator {
+    keys: BTreeMap<Ed25519Public, BTreeSet<NodeKeyValidity>>,
+}
+
+impl MetadataKeyValidator {
+    /// Instantiate a validator with the given node key validity ranges.
+    pub fn new(configs: &[NodeKeyValidity]) -> Self {
+        let mut keys = BTreeMap::<Ed25519Public, BTreeSet<NodeKeyValidity>>::new();
+        for config in configs {
+            keys.entry(config.message_signing_pub_key)
+                .or_default()
+                .insert(config.clone());
+        }
+
+        Self { keys }
+    }
+
+    /// Check whether the given block metadata is valid at the given index.
+    pub fn verify(&self, metadata: &BlockMetadata, block_index: BlockIndex) -> Result<(), Error> {
+        let ranges = self
+            .keys
+            .get(metadata.node_key())
+            .ok_or(Error::UnknownPubKey)?;
+
+        for range in ranges {
+            if range.index_in_range(block_index) {
+                return Ok(metadata.verify()?);
+            }
+        }
+
+        Err(Error::InvalidPubKey)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use mc_blockchain_test_utils::{make_block_id, make_block_metadata_contents};
+    use mc_crypto_keys::{Ed25519Pair, Ed25519Signature};
+    use mc_util_from_random::FromRandom;
+    use mc_util_test_helper::{RngType as FixedRng, SeedableRng};
+
+    fn make_key(seed: u64) -> Ed25519Public {
+        *make_metadata(seed).node_key()
+    }
+
+    fn make_metadata(seed: u64) -> BlockMetadata {
+        let seed_bytes = seed.to_be_bytes();
+        let mut seed = [0u8; 32];
+        seed[0..seed_bytes.len()].copy_from_slice(&seed_bytes);
+        let mut rng = FixedRng::from_seed(seed);
+
+        let signer = Ed25519Pair::from_random(&mut rng);
+
+        BlockMetadata::from_contents_and_keypair(
+            make_block_metadata_contents(make_block_id(&mut rng), &mut rng),
+            &signer,
+        )
+        .expect("BlockMetadata::from_contents_and_keypair")
+    }
+
+    fn make_verifier() -> MetadataKeyValidator {
```

Comment:
nit: should this be named `make_validator`?

Reply:
Good catch! Left over from a rename.